### PR TITLE
Move to Azure Identity

### LINKED ISSUE #1144: Replace Deprecated Auth Providers with Azure Identity

### PR DESCRIPTION
**Pull request by peombwa**
This PR closes #1144 by:
- Replacing deprecated auth providers with Azure Identity's token credentials.
- Updating both unit and Pester tests.

The following improvements have been made to the Authentication module as part of this transition:
- Drops `-AccessToken` from `Invoke-MgGraphRequest`. The goal is to have one authentication/authorization point, `Connect-MgGraph`.
- Drops `-ForceRefresh` from `Connect-MgGraph`. `-ForeceRefresh` is not exposed by Azure Identity as the library will fetch new tokens automatically when new scopes are added.
- Adds a simplified `In-Memory` token cache for process context scope.
- Updates dependencies.

### CODE REVIEW

**Review by timayabi2020 — APPROVED** [review]

**Comment by ddyett** [comment]
do you recall why we added the accesstoken on invoke-mggraphrequest to begin with?

File: src/Authentication/Authentication.Core/Microsoft.Graph.Authentication.Core.csproj [hunk 1]
```diff
@@ -1,22 +1,27 @@
...
-    <PackageReference Include="Microsoft.Graph.Core" Version="1.25.1" />
-    <PackageReference Include="Microsoft.Identity.Client" Version="4.29.0" />
-    <PackageReference Include="Microsoft.IdentityModel.JsonWebTokens" Version="5.6.0" />
-    <PackageReference Include="Microsoft.IdentityModel.Logging" Version="5.6.0" />
-    <PackageReference Include="Microsoft.IdentityModel.Tokens" Version="5.6.0" />
-    <PackageReference Include="Microsoft.Windows.SDK.Contracts" Version="10.0.19041.1" />
-    <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="5.6.0" />
+    <PackageReference Include="Azure.Core" Version="1.22.0" />
+    <PackageReference Include="Azure.Identity" Version="1.5.0" />
+    <PackageReference Include="Microsoft.Graph.Core" Version="2.0.8" />
```
Comment (ddyett) [inline src/Authentication/Authentication.Core/Microsoft.Graph.Authentication.Core.csproj:15]: by the time this goes out wouldn't kiota likely be the release?  Is the plan to convert this later and align with CLI?

**Comment by ddyett** [comment]
can you say more about why the token caches for mac/linux are no longer necessary?

File: src/Authentication/Authentication.Core/TokenCache/InMemoryTokenCacheOptions.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Graph.PowerShell.Authentication.Core.TokenCache
+{
+    public class InMemoryTokenCacheOptions : UnsafeTokenCacheOptions
+    {
+        private readonly ReaderWriterLockSlim _sessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+        internal ReadOnlyMemory<byte> TokenCache { get; private set; }
```
Comment (ddyett) [inline src/Authentication/Authentication.Core/TokenCache/InMemoryTokenCacheOptions.cs:16]: seems odd that the options class is holding the reference to the cache and operating on the object.

Comment (peombwa) [reply src/Authentication/Authentication.Core/TokenCache/InMemoryTokenCacheOptions.cs:16]: Good point! It is a cache manager. It's Azure Identity that named the base class [UnsafeTokenCacheOptions](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.unsafetokencacheoptions?view=azure-dotnet). I opted to use the options suffix to be in sync with Azure Identity naming.

File: src/Authentication/Authentication.Core/Microsoft.Graph.Authentication.Core.csproj (hunk 1 above)
Comment (peombwa) [reply src/Authentication/Authentication.Core/Microsoft.Graph.Authentication.Core.csproj:15]: That's the plan. Swapping the current MS Graph .NET core for Kiota based core should be trivial. I have a work item to integrate dependabot for verion updates - https://github.com/microsoftgraph/msgraph-sdk-powershell/issues/1002.

**Comment by peombwa** [comment]
> do you recall why we added the accesstoken on invoke-mggraphrequest to begin with?

It was because we didn't expose `-AccessToken` on `Connect-MgGraph` at the time.

**Comment by peombwa** [comment]
> can you say more about why the token caches for mac/linux are no longer necessary?

We no longer need them because Azure.Identity securely caches tokens on Windows, MacOS, and Linux (when libsecret-1.so.0 is available). We will fall back to in-memory cache when Azure.Identity can't persist tokens to the underlying OS mechanism ([MsalCachePersistenceException](https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/blob/master/src/Microsoft.Identity.Client.Extensions.Msal/MsalCachePersistenceException.cs) is thrown).

**Comment by peombwa** [comment]
We should verify that successfully sign in and token caching is possible on the following platforms (I'll check them once verified):
- Windows:
  - [ ] PS 5.1
  - [ ] PS 7
  - [ ] WSL
- MacOS:
  - [ ] Native PS core session via terminal
  - [ ] Docker
- Linux:
  - [ ] Cloud Shell
  - [ ] Ubuntu Server

**Comment by georgend** [comment]
> do you recall why we added the accesstoken on invoke-mggraphrequest to begin with?

We had added it for BYOT scenarios. For example authenticating against MSI, getting a token and (re)-using it with the SDK.

**Review by georgend — APPROVED** [review]

**Pull request merged**